File: verticalTraversal.cpp
/*
Intuition: First we will keep a track of the horizontal distance for each node.
For example: root node will have a horizontal distance of 0. its left node will have a HD of -1
and right node will have a HD of 1.
Level Order will be the easiest. 

Time Complexity: O(N), N = Number of nodes.
Space Complexity: O(N), N = Number of nodes.

*/
class Solution {
public:
    vector<vector<int>> verticalOrder(TreeNode* root) {
        vector<vector<int>> result;
        if ( root == NULL) return result;

        map<int, vector<int>> map;
        queue<pair<TreeNode*,int>>queue;
        queue.push({root, 0});
        int minHD = 0, maxHD = 0;
        while ( queue.size() != 0){
            
            auto currPair = queue.front();
            queue.pop();
            
            TreeNode* currNode = currPair.first;
            int HD = currPair.second;
            
            minHD = min(minHD, HD);
            maxHD = max(maxHD, HD);
            if ( map.find(HD) == map.end()){
                vector<int> partialResult;
                map[HD] = partialResult;
            }
            map[HD].push_back(currNode->val);
            
            if ( currNode->left != NULL) queue.push({currNode->left, HD-1});
            if ( currNode->right != NULL) queue.push({currNode->right, HD+1});

        }
        for ( int i = minHD; i <=maxHD; i++){
            result.push_back(map[i]);
        }
        return result;
    }
};

File: verticalTraversal.java
import java.util.*;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int val) { this.val = val; }
}

class Solution {
    public List<List<Integer>> verticalOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        if (root == null) {
            return result;
        }
        
        Map<Integer, List<Integer>> columnMap = new TreeMap<>();
        Queue<TreeNode> queue = new LinkedList<>();
        Queue<Integer> columnQueue = new LinkedList<>();
        
        queue.offer(root);
        columnQueue.offer(0);
        
        while (!queue.isEmpty()) {
            TreeNode node = queue.poll();
            int column = columnQueue.poll();
            
            // Add the current node's value to the column's list
            columnMap.computeIfAbsent(column, k -> new ArrayList<>()).add(node.val);
            
            // Process left child
            if (node.left != null) {
                queue.offer(node.left);
                columnQueue.offer(column - 1);
            }
            
            // Process right child
            if (node.right != null) {
                queue.offer(node.right);
                columnQueue.offer(column + 1);
            }
        }
        
        // Convert the column map to the result list
        result.addAll(columnMap.values());
        
        return result;
    }
}

